File: delightreading-server/src/test/java/com/delightreading/user/UserProfileRepositoryIT.java
package com.delightreading.user;

import com.delightreading.SpringApplicationContextUtil;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJsonTesters;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@DataJpaTest
@AutoConfigureJsonTesters
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@ContextConfiguration(classes = {SpringApplicationContextUtil.class})
@EnableAutoConfiguration
public class UserProfileRepositoryIT {

    @Autowired
    UserProfileRepository userProfileRepository;

    @Autowired
    private TestEntityManager entityManager;

    @Before
    public void setup() {
        var account1 = UserAccountEntityRepositoryIT.buildEntity("TEST-UserUID1", "TEST-Username1", "TEST-givenName1", Arrays.asList("email1a@test.com", "email1b@test.com"));
        entityManager.persistAndFlush(account1);
        var profile1 = UserProfileRepositoryIT.buildEntity(Arrays.asList(Experience.builder().kind("exp-kind1").title("exp-title1").build()), Arrays.asList("math", "science"), "male", account1);
        entityManager.persistAndFlush(profile1);

        var account2 = UserAccountEntityRepositoryIT.buildEntity(null, "TEST-Username2", "TEST-givenName2", Arrays.asList("email2a@test.com", "email2b@test.com"));
        entityManager.persistAndFlush(account2);
        var profile2 = UserProfileRepositoryIT.buildEntity(Arrays.asList(Experience.builder().kind("exp-kind2").title("exp-title2").build()), Arrays.asList("literature", "science"), "female", account2);
        entityManager.persistAndFlush(profile2);
    }

    @Test
    public void crud_simple() {

        //var match = userProfileRepository.findByGender("male").get(0);
        var match = userProfileRepository.findByAccountUsername("TEST-Username1").get();

        assertThat (match).isNotNull();
        assertThat(match.getGender()).isEqualTo("male");
        assertThat(match.getAccount()).isNotNull();
        assertThat(match.getAccount().getUsername()).isEqualTo("TEST-Username1");
        assertThat(match.getInterests()).containsExactlyInAnyOrder("math", "science");
        assertThat(match.getEducation()).hasSize(1);
        assertThat(match.getEducation().get(0).getKind()).isEqualTo("exp-kind1");
    }

    public static UserProfileEntity buildEntity(List<Experience> education, List<String> interests, String gender, UserAccountEntity account) {
        UserProfileEntity authEntity = UserProfileEntity.builder()
                .education(education)
                .interests(interests)
                .gender(gender)
                .account(account)
                .build();

        return authEntity;
    }
}

File: delightreading-backend/src/test/kotlin/com/delightreading/user/UserProfileRepositoryIT.kt
package com.delightreading.user

import com.delightreading.SpringApplicationContextUtil
import com.delightreading.user.model.Experience
import com.delightreading.user.model.UserAccountEntity
import com.delightreading.user.model.UserProfileEntity
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJsonTesters
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.*
import javax.transaction.Transactional

@ExtendWith(SpringExtension::class)
@Transactional
@SpringBootTest(
    classes = arrayOf(
        TestEntityManager::class,
        SpringApplicationContextUtil::class
    )
)
@EnableAutoConfiguration
class UserProfileRepositoryIT(
    @Autowired val entityManager: TestEntityManager,
    @Autowired val userProfileRepository: UserProfileRepository

) {

    @BeforeEach
    fun setup() {
        val account1 = UserAccountEntity(
            uid = "TEST-UserUID1",
            username = "TEST-Username1",
            givenName = "TEST-givenName1",
            emails = Arrays.asList("email1a@test.com", "email1b@test.com")
        )
        entityManager.persistAndFlush(account1)
        val profile1 = UserProfileEntity(
            education =  Arrays.asList(
                Experience(kind = "exp-kind1", title = "exp-title1")
            ),
            interests = Arrays.asList("math", "science"),
            gender = "male",
            account = account1
        )
        entityManager.persistAndFlush<UserProfileEntity>(profile1)

        val account2 = UserAccountEntity(
            uid = null,
            username = "TEST-Username2",
            givenName = "TEST-givenName2",
            emails = Arrays.asList("email2a@test.com", "email2b@test.com")
        )
        entityManager.persistAndFlush(account2)
        val profile2 = UserProfileEntity(
            education =  Arrays.asList(Experience(kind = "exp-kind2", title = "exp-title2")),
            interests = Arrays.asList("literature", "science"),
            gender = "female",
            account = account2
        )
        entityManager.persistAndFlush<UserProfileEntity>(profile2)
    }

    @Test
    fun crud_simple() {

        //var match = userProfileRepository.findByGender("male").get(0);
        val match = userProfileRepository.findByAccountUsername("TEST-Username1")

        assertThat(match).isNotNull()
        assertThat(match!!.gender).isEqualTo("male")
        assertThat(match!!.account).isNotNull()
        assertThat(match!!.account!!.username).isEqualTo("TEST-Username1")
        assertThat(match!!.interests).containsExactlyInAnyOrder("math", "science")
        assertThat(match!!.education).hasSize(1)
        assertThat(match!!.education!![0].kind).isEqualTo("exp-kind1")
    }

}